File: node/src/domains/api_keys.js
// api_keys — issue/list/verify/rotate/revoke API keys with scopes; every key records created_at (seconds, the core
// clock seam, preserved across rotate). (1) NO PLAINTEXT AT REST: the secret is shown
// ONCE; only sha256(secret) (the digest part) is stored. (2) CONSTANT-TIME, NON-ENUMERABLE VERIFY: every
// verify hashes and runs ONE timingSafeEqual against a record (a dummy when the id is unknown) — an unknown id
// and a wrong secret are the same {valid:false} after the same work; scopes only when valid. (3) ROTATION
// invalidates the old secret. (4) REVOCATION is monotonic. Key is `ak_<id>_<secret>`; the prefix is public,
// the secret is not. Store names and shapes match python/go; durable.
//
// OWNERSHIP — a key is USER-SCOPED: it belongs to the caller who created it. create/get/list/rotate/revoke call
// requireIdentity (the core seam) FIRST, the OWNER is stamped from the authenticated subject at create (never a body
// field), and a management op on another caller's key id is 404 — byte-identical to a missing id, so the enumerable
// sequential id leaks no existence (the tenancy not-yours==not-found pattern). apiKeysList is the same owner-scoping
// over a COLLECTION (only the caller's keys, paginated, secret/owner-blind; a stranger gets an empty page, never 403).
// apiKeysVerify stays PUBLIC (see its declaration).
import { randomBytes, timingSafeEqual } from 'node:crypto';

import { intParam, nextId, problem, requireIdentity, sendJSON, storeGet, storePut, storeValues, testNow } from '../core/runtime.js';
import { digestHex } from '../parts/digest.js';
import { paginate } from '../parts/paginate.js';
import { isWellFormed } from '../parts/well_formed.js';

const DUMMY_HASH = digestHex('api_keys_absent_record_filler');
// state in store: seq "api_keys_key" · ns "api_keys_records" String(id) -> {id,name,owner,scopes,prefix,secret_hash,status}

// the public view NEVER includes secret_hash OR owner (owner is private, like the hash); created_at IS public
const publicView = (rec) => ({ id: rec.id, name: rec.name, scopes: rec.scopes, prefix: rec.prefix, status: rec.status, created_at: rec.created_at });

async function issue(rec) {
  const secret = randomBytes(24).toString('base64url');
  const stored = { ...rec, secret_hash: digestHex(secret) };
  await storePut('api_keys_records', String(rec.id), stored);
  return { ...publicView(stored), key: `ak_${rec.id}_${secret}` };
}

// load expects requireIdentity to have run already; it does the path-int check, loads, then enforces owner==caller,
// returning 404 for a missing OR cross-owner id (not-yours == not-found: existence never leaks cross-owner).
async function load(req, res, params, owner) {
  const kid = intParam(params.key_id);
  if (kid === null) { problem(res, 422, 'invalid key id'); return null; }
  const rec = await storeGet('api_keys_records', String(kid));
  if (rec === undefined || rec.owner !== owner) { problem(res, 404, 'api key not found'); return null; }
  return rec;
}

export async function apiKeysCreate(req, res, params, body) {
  const owner = await requireIdentity(req, res); // the runtime parsed the body already, so identity-first is fine here
  if (owner === null) return;
  if (!body || !isWellFormed(body.name) || !Array.isArray(body.scopes) || body.scopes.length === 0
      || !body.scopes.every(isWellFormed)) {
    return problem(res, 422, 'invalid body');
  }
  const kid = await nextId('api_keys_key');
  // owner derived from the token, never client-set; created_at = the birth time via the core clock seam (preserved across rotate)
  const at = testNow(req);
  sendJSON(res, 201, await issue({ id: kid, name: body.name, owner, scopes: body.scopes, prefix: `ak_${kid}`, status: 'active', created_at: at }));
}

export async function apiKeysGet(req, res, params) {
  const owner = await requireIdentity(req, res);
  if (owner === null) return;
  const rec = await load(req, res, params, owner);
  if (rec !== null) sendJSON(res, 200, publicView(rec));
}

export async function apiKeysList(req, res) {
  const owner = await requireIdentity(req, res);
  if (owner === null) return;
  // OWNER-SCOPED LIST: only the caller's keys leave the store — filtered INLINE on the authenticated owner, mapped to
  // the secret/owner-blind publicView, then a BOUNDED page. Cross-owner isolation proven by I8; a stranger gets an
  // empty page, never 403.
  // sorted by id (the stable order) — rotate/revoke re-write a row + bump its rowid, so an explicit id-sort is
  // required for a stable paged walk (the notifications/admin precedent; tenancy/audit_log never UPDATE a row).
  const mine = (await storeValues('api_keys_records')).filter((r) => r.owner === owner)
    .sort((a, b) => a.id - b.id).map(publicView);
  const q = new URL(req.url, 'http://localhost').searchParams;
  const { items, next, ok } = paginate(mine, q.get('cursor') || '', q.get('limit') || '');
  if (!ok) return problem(res, 422, 'invalid cursor or limit');
  sendJSON(res, 200, { results: items, next_cursor: next });
}

export async function apiKeysVerify(req, res, params, body) {
  // mutation-auth: public — INTENTIONALLY unauthenticated. The `ak_<id>_<secret>` key IS the credential (like a
  // login): a caller verifies it BEFORE it has a session, so requireIdentity would break the route's purpose. It
  // mutates no stored state on behalf of any user — it only recomputes the hash and runs one constant-time compare.
  // The owner-scoping guards the MANAGEMENT ops (create/get/rotate/revoke), not this credential check.
  if (!body || typeof body.key !== 'string' || body.key === '') return problem(res, 422, 'invalid body');
  // parse `ak_<id>_<secret>`; a malformed key still runs the dummy compare (uniform timing)
  let keyId = '';
  let secret = '';
  const firstUnderscore = body.key.indexOf('_');
  const secondUnderscore = body.key.indexOf('_', firstUnderscore + 1);
  if (body.key.startsWith('ak_') && secondUnderscore > firstUnderscore) {
    keyId = body.key.slice(3, secondUnderscore);
    secret = body.key.slice(secondUnderscore + 1);
  }
  const rec = keyId ? await storeGet('api_keys_records', keyId) : undefined;
  const stored = rec ? rec.secret_hash : DUMMY_HASH;
  // ALWAYS one constant-time compare (hex strings are fixed length 64, so timingSafeEqual is safe)
  const got = Buffer.from(digestHex(secret));
  const want = Buffer.from(stored);
  const match = got.length === want.length && timingSafeEqual(got, want);
  const valid = rec !== undefined && rec.status === 'active' && match;
  sendJSON(res, 200, { valid, scopes: valid ? rec.scopes : [] });
}

export async function apiKeysRotate(req, res, params) {
  const owner = await requireIdentity(req, res);
  if (owner === null) return;
  const rec = await load(req, res, params, owner);
  if (rec === null) return;
  sendJSON(res, 200, await issue(rec)); // new secret + hash replaces the old; the old can never verify again
}

export async function apiKeysRevoke(req, res, params) {
  const owner = await requireIdentity(req, res);
  if (owner === null) return;
  const rec = await load(req, res, params, owner);
  if (rec === null) return;
  const revoked = { ...rec, status: 'revoked' }; // monotonic + idempotent: revoked is TERMINAL
  await storePut('api_keys_records', String(rec.id), revoked);
  sendJSON(res, 200, publicView(revoked));
}

File: go/internal/domains/api_keys/api_keys.go
// Package api_keys — issue/list/verify/rotate/revoke API keys with scopes; every key records created_at (seconds,
// the core clock seam, preserved across rotate). (1) NO PLAINTEXT AT REST: the secret is
// shown ONCE; only sha256(secret) (the digest part) is stored. (2) CONSTANT-TIME, NON-ENUMERABLE VERIFY: every
// verify hashes and runs ONE hmac.Equal against a record (a dummy when the id is unknown) — an unknown id and a
// wrong secret are the same {valid:false} after the same work; scopes only when valid. (3) ROTATION invalidates
// the old secret. (4) REVOCATION is monotonic. Key is `ak_<id>_<secret>`; the prefix is public, the secret is
// not. Store names and shapes match python/node; durable.
//
// OWNERSHIP — a key is USER-SCOPED: it belongs to the caller who created it. ApiKeysCreate/Get/List/Rotate/Revoke require
// core.RequireIdentity, the OWNER is stamped from the authenticated subject at create (never a body field), and a
// management op on another caller's key id is 404 — byte-identical to a missing id, so the enumerable sequential
// id leaks no existence (the tenancy not-yours==not-found pattern). ApiKeysList is the same owner-scoping over a
// COLLECTION (only the caller's keys, paginated, secret/owner-blind; a stranger gets an empty page, never 403). Precedence ×3: create is body-only so DecodeJSON
// runs FIRST (drain + 413), then RequireIdentity, then validation; the {key_id} routes run RequireIdentity FIRST so
// a no-token probe is 401 before the path-422. ApiKeysVerify stays PUBLIC (see its `mutation-auth: public`).
package api_keys

import (
	"crypto/hmac"
	"crypto/rand"
	"encoding/base64"
	"net/http"
	"sort"
	"strconv"
	"strings"

	"app/internal/core"
	"app/internal/parts/digest"
	"app/internal/parts/paginate"
	"app/internal/parts/well_formed"
)

var apiKeysDummyHash = digest.DigestHex("api_keys_absent_record_filler")

type apiKeysRecord struct {
	Id         int      `json:"id"`
	Name       string   `json:"name"`
	Owner      string   `json:"owner"`
	Scopes     []string `json:"scopes"`
	Prefix     string   `json:"prefix"`
	SecretHash string   `json:"secret_hash"`
	Status     string   `json:"status"`
	CreatedAt  int      `json:"created_at"`
}

var apiKeysRecords = core.NewKV[string, apiKeysRecord]("api_keys_records")

func apiKeysPublic(rec apiKeysRecord) map[string]any {
	return map[string]any{"id": rec.Id, "name": rec.Name, "scopes": rec.Scopes, "prefix": rec.Prefix,
		"status": rec.Status, "created_at": rec.CreatedAt} // NEVER secret_hash OR owner (owner private); created_at public
}

func apiKeysIssue(rec apiKeysRecord) map[string]any {
	buf := make([]byte, 24)
	if _, err := rand.Read(buf); err != nil {
		panic("api_keys: randomness unavailable") // a guessable secret is worse than a crash
	}
	secret := base64.RawURLEncoding.EncodeToString(buf)
	rec.SecretHash = digest.DigestHex(secret)
	apiKeysRecords.Set(strconv.Itoa(rec.Id), rec)
	out := apiKeysPublic(rec)
	out["key"] = "ak_" + strconv.Itoa(rec.Id) + "_" + secret
	return out
}

func ApiKeysCreate(w http.ResponseWriter, r *http.Request) {
	// body-only POST: DecodeJSON FIRST (drain the stream + enforce the 413 cap), THEN RequireIdentity, THEN validate
	// — so an ill-typed/oversize body is a 413/422 before auth, and a no-token caller is 401, identical ×3.
	in, ok := core.DecodeJSON[struct {
		Name   *string  `json:"name"`
		Scopes []string `json:"scopes"`
	}](w, r)
	if !ok {
		return
	}
	owner, ok := core.RequireIdentity(w, r)
	if !ok {
		return
	}
	if in.Name == nil || !well_formed.IsWellFormed(*in.Name) || len(in.Scopes) == 0 {
		core.WriteProblem(w, 422, "invalid body")
		return
	}
	for _, s := range in.Scopes {
		if !well_formed.IsWellFormed(s) {
			core.WriteProblem(w, 422, "invalid body")
			return
		}
	}
	kid := core.NextID("api_keys_key")
	// owner derived from the token, never client-set; created_at = the birth time via the core clock seam (preserved across rotate)
	rec := apiKeysRecord{Id: kid, Name: *in.Name, Owner: owner, Scopes: in.Scopes, Prefix: "ak_" + strconv.Itoa(kid),
		Status: "active", CreatedAt: int(core.TestNow(r))}
	core.WriteJSON(w, 201, apiKeysIssue(rec))
}

// apiKeysLoad expects RequireIdentity to have run already (so a no-token probe is 401 before the path-422); it does
// the path-int check, then loads, then enforces owner==caller, returning 404 for a missing OR cross-owner id.
func apiKeysLoad(w http.ResponseWriter, r *http.Request, owner string) (apiKeysRecord, bool) {
	kid, err := strconv.Atoi(r.PathValue("key_id"))
	if err != nil {
		core.WriteProblem(w, 422, "invalid key id")
		return apiKeysRecord{}, false
	}
	rec, exists := apiKeysRecords.Get(strconv.Itoa(kid))
	if !exists || rec.Owner != owner {
		core.WriteProblem(w, 404, "api key not found") // not-yours == not-found: existence never leaks cross-owner
		return apiKeysRecord{}, false
	}
	return rec, true
}

func ApiKeysGet(w http.ResponseWriter, r *http.Request) {
	owner, ok := core.RequireIdentity(w, r) // AUTH first: a no-token probe is 401 before the path-422
	if !ok {
		return
	}
	if rec, ok := apiKeysLoad(w, r, owner); ok {
		core.WriteJSON(w, 200, apiKeysPublic(rec))
	}
}

func ApiKeysList(w http.ResponseWriter, r *http.Request) {
	owner, ok := core.RequireIdentity(w, r)
	if !ok {
		return
	}
	// OWNER-SCOPED LIST: only the caller's keys leave the store — filtered INLINE on the authenticated owner, mapped
	// to the secret/owner-blind public view, then a BOUNDED page. Cross-owner isolation proven by I8; a stranger
	// gets an empty page, never 403.
	mine := make([]apiKeysRecord, 0)
	for _, rec := range apiKeysRecords.All() {
		if rec.Owner == owner {
			mine = append(mine, rec)
		}
	}
	// sorted by id (the stable order) — rotate/revoke re-write a row + bump its rowid, so an explicit id-sort is
	// required for a stable paged walk (the notifications/admin precedent; tenancy/audit_log never UPDATE a row).
	sort.Slice(mine, func(i, j int) bool { return mine[i].Id < mine[j].Id })
	views := make([]map[string]any, len(mine))
	for i, rec := range mine {
		views[i] = apiKeysPublic(rec)
	}
	q := r.URL.Query()
	page, next, valid := paginate.Paginate(views, q.Get("cursor"), q.Get("limit"))
	if !valid {
		core.WriteProblem(w, 422, "invalid cursor or limit")
		return
	}
	var nc any
	if next != "" {
		nc = next
	}
	core.WriteJSON(w, 200, map[string]any{"results": page, "next_cursor": nc})
}

func ApiKeysVerify(w http.ResponseWriter, r *http.Request) {
	// mutation-auth: public — INTENTIONALLY unauthenticated. The `ak_<id>_<secret>` key IS the credential (like a
	// login): a caller verifies it BEFORE it has a session, so RequireIdentity would break the route's purpose. It
	// mutates no stored state on behalf of any user — it only recomputes the hash and runs one constant-time
	// compare. The owner-scoping guards the MANAGEMENT ops (create/get/rotate/revoke), not this credential check.
	in, ok := core.DecodeJSON[struct {
		Key *string `json:"key"`
	}](w, r)
	if !ok {
		return
	}
	if in.Key == nil || *in.Key == "" {
		core.WriteProblem(w, 422, "invalid body")
		return
	}
	// parse `ak_<id>_<secret>`; a malformed key still runs the dummy compare (uniform timing)
	keyID, secret := "", ""
	parts := strings.SplitN(*in.Key, "_", 3)
	if len(parts) == 3 && parts[0] == "ak" {
		keyID, secret = parts[1], parts[2]
	}
	rec, found := apiKeysRecords.Get(keyID)
	stored := apiKeysDummyHash
	if found {
		stored = rec.SecretHash
	}
	match := hmac.Equal([]byte(digest.DigestHex(secret)), []byte(stored)) // ALWAYS one constant-time compare
	valid := found && rec.Status == "active" && match
	scopes := []string{}
	if valid {
		scopes = rec.Scopes
	}
	core.WriteJSON(w, 200, map[string]any{"valid": valid, "scopes": scopes})
}

func ApiKeysRotate(w http.ResponseWriter, r *http.Request) {
	owner, ok := core.RequireIdentity(w, r) // AUTH first: a no-token probe is 401 before the path-422
	if !ok {
		return
	}
	rec, ok := apiKeysLoad(w, r, owner)
	if !ok {
		return
	}
	core.WriteJSON(w, 200, apiKeysIssue(rec)) // new secret + hash replaces the old; the old can never verify again
}

func ApiKeysRevoke(w http.ResponseWriter, r *http.Request) {
	owner, ok := core.RequireIdentity(w, r) // AUTH first: a no-token probe is 401 before the path-422
	if !ok {
		return
	}
	rec, ok := apiKeysLoad(w, r, owner)
	if !ok {
		return
	}
	rec.Status = "revoked" // monotonic + idempotent: revoked is TERMINAL
	apiKeysRecords.Set(strconv.Itoa(rec.Id), rec)
	core.WriteJSON(w, 200, apiKeysPublic(rec))
}
